include/dslash_helper.cuh

```diff
@@ -12,4 +18,359 @@ namespace quda {
+
+  template <typename Float>
+  std::ostream& operator<<(std::ostream& out, const DslashArg<Float> &arg)
+  {
+    out << "parity = " << arg.parity << std::endl;
+    out << "nParity = " << arg.nParity << std::endl;
+    out << "nFace = " << arg.nFace << std::endl;
+    out << "reconstruct = " << arg.reconstruct << std::endl;
+    out << "X0h = " << arg.X0h << std::endl;
+    out << "dim = { "; for (int i=0; i<5; i++) out << arg.dim[i] << (i<4 ? ", " : " }"); out << std::endl;
+    out << "commDim = { "; for (int i=0; i<4; i++) out << arg.commDim[i] << (i<3 ? ", " : " }"); out << std::endl;
+    out << "volumeCB = " << arg.volumeCB << std::endl;
+    out << "kappa = " << arg.kappa << std::endl;
+    out << "dagger = " << arg.dagger << std::endl;
+    out << "xpay = " << arg.xpay << std::endl;
+    out << "kernel_type = " << arg.kernel_type << std::endl;
+    out << "remote_write = " << arg.remote_write << std::endl;
+    out << "threads = " << arg.threads << std::endl;
+    out << "threadDimMapLower = { ";
+    for (int i=0; i<4; i++) out << arg.threadDimMapLower[i] << (i<3 ? ", " : " }"); out << std::endl;
+    out << "threadDimMapUpper = { ";
+    for (int i=0; i<4; i++) out << arg.threadDimMapUpper[i] << (i<3 ? ", " : " }"); out << std::endl;
+    return out;
+  }
+
+  //static declarations
+  static bool init = false;
+  static char ghost_str[TuneKey::aux_n]; // string with ghostDim information
+
+  template <typename Float>
+  class Dslash : public TunableVectorY {
+
+  protected:
+
+    DslashArg<Float> &arg;
+    const ColorSpinorField &in;
+
+    const int nDimComms;
+
+    char aux_base[TuneKey::aux_n];
+    char aux[8][TuneKey::aux_n];
+
+    /**
+       @brief Set the base strings used by the different dslash kernel
+       types for autotuning.
+    */
+    inline void fillAuxBase() {
+      char comm[5];
+      comm[0] = (arg.commDim[0] ? '1' : '0');
+      comm[1] = (arg.commDim[1] ? '1' : '0');
+      comm[2] = (arg.commDim[2] ? '1' : '0');
+      comm[3] = (arg.commDim[3] ? '1' : '0');
+      comm[4] = '\0';
+      strcpy(aux_base,",comm=");
+      strcat(aux_base,comm);
+
+      if (arg.xpay) strcat(aux_base,",xpay");
+      if (arg.dagger) strcat(aux_base,",dagger");
+    }
+
+    /**
+       @brief Specialize the auxiliary strings for each kernel type
+       @param[in] kernel_type The kernel_type we are generating the string got
+       @param[in] kernel_str String corresponding to the kernel type
+    */
+    inline void fillAux(KernelType kernel_type, const char *kernel_str) {
+      strcpy(aux[kernel_type],kernel_str);
+      if (kernel_type == INTERIOR_KERNEL) strcat(aux[kernel_type],ghost_str);
+      strcat(aux[kernel_type],aux_base);
+    }
+
+    bool tuneGridDim() const { return false; }
+    unsigned int minThreads() const { return arg.threads; }
+
+    template <typename T, typename Arg>
+    inline void launch(T *f, const TuneParam &tp, Arg &arg, const cudaStream_t &stream) {
+      if (1) { // test on Volta
```

Comment:
Was this tested?

Reply:
This code has been gone for a while, so not sure why you're looking at this out of date file.